jormungandr/src/watch_client.rs

```diff
@@ -0,0 +1,329 @@
+        // the other is when the receiver is lagging.
+        let live_stream: SubscriptionTryStream<BroadcastStream<Block>> =
+            block_receiver.map_err(|e| chain_network::error::Error::new(Code::Aborted, e));
+
+        Ok(live_stream)
+    }
+
+    #[instrument(skip(self))]
+    async fn tip_subscription(
+        &self,
+    ) -> Result<Self::TipSubscriptionStream, chain_network::error::Error> {
+        let tip_receiver: SubscriptionStream<_> = WatchStream::new(self.tip_receiver.clone())
+            .map::<Result<Header, chain_network::error::Error>, _>(Ok);
+
+        Ok(tip_receiver)
+    }
+
+    #[instrument(skip(self))]
+    async fn sync_multiverse(
+        &self,
+        from: BlockIds,
+    ) -> Result<Self::SyncMultiverseStream, chain_network::error::Error> {
+        let (handle, future) = intercom::stream_reply(32);
+
+        self.request_tx
+            .clone()
+            .send(RequestMsg::SyncMultiverse { from, handle })
+            .await
+            .map_err(|e| chain_network::error::Error::new(Code::Unavailable, e))?;
+
+        let stream = future
+            .await
+            .map_err(|e| chain_network::error::Error::new(Code::Internal, e))?;
+
+        Ok(stream.map_err(|e| chain_network::error::Error::new(Code::Internal, e)))
+    }
+}
+
+async fn handle_sync_multiverse(
+    checkpoints: BlockIds,
+    blockchain: &Blockchain,
+    storage: &Storage,
+    sink: &mut intercom::ReplyStreamSink<Block>,
+) -> Result<(), intercom::Error> {
+    let mut checkpoints = checkpoints
+        .iter()
+        .map(|id| {
+            let deser = HeaderHash::deserialize(id.as_bytes())
+                .map_err(intercom::Error::invalid_argument)?;
+
+            let chain_length = storage
+                .get_chain_length(deser)
+                .ok_or_else(|| intercom::Error::invalid_argument("unknown checkpoint"))?;
+
+            Ok::<_, intercom::Error>((chain_length, deser))
+        })
+        .collect::<Result<Vec<_>, _>>()?;
+
+    let block0 = blockchain.block0();
+
+    // we are adding 1 to the lsb later, because it is present then we don't need to send it.
+    // but if the checkpoints are empty then it means the last stable block is the block0, and we
+    // send it here.
+    if checkpoints.is_empty() {
+        let block = storage
+            .get(*block0)
+            .map_err(intercom::Error::failed)
+            .and_then(|maybe_block0| {
+                maybe_block0.ok_or_else(|| intercom::Error::failed("block0 not found in storage"))
+            })?;
+
+        sink.send(Ok(chain_network::data::Block::from_bytes(
+            block.serialize_as_vec().unwrap(),
+        )))
+        .await
+        .map_err(intercom::Error::failed)?;
+    }
+
+    checkpoints.sort_unstable();
+
+    // TODO: not sure if this is always true
+    const BLOCK0_CHAIN_LENGTH: u32 = 0;
+    let (lsb_chain_length, lsb) = checkpoints.pop().unwrap_or((BLOCK0_CHAIN_LENGTH, *block0));
```

Comment:
`pop` works the opposite way as you would like here AFAIU. It will remove the last block, which is the one furthest down the chain, not the lsb